### ISSUE TITLE
software cache for OpenCL kernel

### ISSUE THREAD

**Issue by wangliwei-intel** [openvinotoolkit/openvino#907]
Hi,
I'm running openvino's demos on GPU, I found each time all kernels are compiled from scratch, so
I have to wait quite long time.
Is there any kind of software cache through which the applications can start faster next time?

B. R.

**Comment by alalek** [comment]
Intel NEO OpenCL runtime provides experimental cache support.
The latest discussion is here: https://github.com/intel/compute-runtime/issues/310 (with link to actual documentation)

**Comment by brmarkus** [comment]
Either export an ENV variable "`cl_cache_dir`" which points to a (writable and accessible directory) or create a directory "`cl_cache`" in the current working directory where the application (or e.g. Python script) gets executed.

**Comment by wangliwei-intel** [comment]
Thank your information, I'm gonna give a try.
By the way, Does cldnn support multi-task compilation? I found it compiles kernel one by one.

**Comment by wangliwei-intel** [comment]
Hi,
Will clDNN support multi-task compilation?

B. R.

**Comment by JulienMaille** [comment]
@wangliwei-intel any update on this? By "long-time" do you mean 20+sec for each kernel?

**Comment by vladimir-paramuzov** [comment]
@wangliwei-intel  Software cache can be enabled via [cl_cache feature](https://github.com/intel/compute-runtime/blob/master/opencl/doc/FAQ.md#feature-cl_cache) in ocl runtime or via `CACHE_DIR` plugin [option](https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/include/ie_plugin_config.hpp#L376) (recommended). 
As for compilation, currently cldnn groups multiple kernels into batches (~10 kernels per batch) and compile them sequentially. After merging #3598 compilation of the batches will be done in parallel.

If you have any other questions, feel free to reopen / create new issue.

**Comment by natelowry** [comment]
note for anyone else here from the future.  we updated some packages and the `plugins.xml` setting stopped working AND the `cl_cache_dir` environment variable stopped working.  we were able to get it working by creating the `cl_cache` directory where the executable lives per @brmarkus's suggestion <3

**Comment by JulienMaille** [comment]
> note for anyone else here from the future.  we updated some packages and the `plugins.xml` setting stopped working AND the `cl_cache_dir` environment variable stopped working.  we were able to get it working by creating the `cl_cache` directory where the executable lives per @brmarkus's suggestion <3

What do you mean? When did it stop working?

**Comment by natelowry** [comment]
we updated the ML.NET runtime and switched to dotnet core 7 (from .NET Framework 4.8), but didn't update the runtime.  in the past the plugins.xml cache directory was getting picked up, but something in that transition of runtimes made that not the case anymore.  my guess is somewhere in that code they're setting the directory explicitly, but i haven't had time to check.

just wanted to post here in case someone found this thread and needed confirmation that the `cl_cache` directory will work as a last-ditch effort.
